Aaron/issues/translation

aaronskiba opened the pull request:
- Enable missing French translations: #404
- Remove duplicate "Sign Up" and "Help" messages in "About Us": #407
- Fix broken "Help" link: #412

aaronskiba committed:
Edit string to match translation.io entry

This addresses "About Us" bug mentioned in issue #404

aaronskiba committed:
Change "_" to " " for output_types

Enables proper export to translation.io

aaronskiba committed:
Add access enums for translation.io sync

aaronskiba committed:
Format strings for translation.io sync

aaronskiba committed:
Add translation of db field

Need this for translation "Research domain" dropdown elements in "Project Details" panel of Plan

aaronskiba committed:
Comment out repeated Sign-Up & Help text

This addresses issue #407

aaronskiba committed:
Update CHANGELOG.md

aaronskiba committed:
Address rubocop offense

aaronskiba commented on app/presenters/research_output_presenter.rb:
Maybe refactoring is needed for the array in `def self.selectable_subjects`? If the elements didn't contain commas and other symbols, I would've created `enum selectable_subject` in `class ResearchOutput`. Also, I'm not sure if the prepended `XX-` can be removed from each of the array elements.

aaronskiba commented on app/views/static_pages/about_us.html.erb:
This remedies issue #407. No translations are currently available for the "Sign up" and "Help" messages I commented out here. Otherwise, I probably would've kept them and commented out the other "Sign up" and "Help" messages.

lagoan commented on app/presenters/research_output_presenter.rb:
```diff
@@ -11,13 +11,13 @@ def initialize(research_output:)
   def selectable_output_types
     ResearchOutput.output_types
-                  .map { |k, _v| [k.humanize, k] }
+                  .map { |k, _v| [_(k.humanize), k] }
```
This wont work as the key is not yet known when the translation rake task analyses the code. We should set it back to ```.map { |k, _v| [k.humanize, k] }``` to avoid confusion.

same as the change on line 20.

aaronskiba replied:
I think we might be ok here because of the way we are listing the strings of the enums in `_research_output.erb`.
I added a couple of translations in dmp-assistant-test. "Audiovisual" is an entry for `ResearchOutput.output_types` and "Restricted" is an entry for `ResearchOutput.accesses` (they both got exported to translation.io after being defined in `_research_output.erb`). The translations are looking ok in the dropdowns.

<img width="787" alt="Screenshot 2023-08-01 at 3 22 32 PM" src="https://github.com/portagenetwork/roadmap/assets/71047780/49459792-436a-4542-8700-b063ec451f7b">


<img width="317" alt="Screenshot 2023-08-01 at 3 21 37 PM" src="https://github.com/portagenetwork/roadmap/assets/71047780/5dd98648-df50-4608-a407-eae4e81c9748">


<img width="256" alt="Screenshot 2023-08-01 at 3 21 55 PM" src="https://github.com/portagenetwork/roadmap/assets/71047780/9af6718c-5b9d-492c-be86-e0c0ad6acba2">

aaronskiba replied:
Or sorry, maybe i'm misunderstanding your concern. If a translation exists, then `_(k.humanize)` will provide the translation. However, if `k.humanize` is not already exported to translation.io, then `_(k.humanize)` cannot help us.

lagoan commented on app/presenters/research_output_presenter.rb:
```diff
@@ -54,22 +54,23 @@ def byte_sizable?
-      '32-Physics',
-      '12-Social and Behavioural Sciences',
-      '42-Thermal Engineering/Process Engineering'
+      "23-#{_('Agriculture, Forestry, Horticulture and Veterinary Medicine')}",
+      "21-#{_('Biology')}",
+      "31-#{_('Chemistry')}",
+      "44-#{_('Computer Science, Electrical and System Engineering')}",
+      "45-#{_('Construction Engineering and Architecture')}",
+      "34-#{_('Geosciences (including Geography)')}",
+      "11-#{_('Humanities')}",
+      "43-#{_('Materials Science and Engineering')}",
+      "33-#{_('Mathematics')}",
+      "41-#{_('Mechanical and industrial Engineering')}",
+      "22-#{_('Medicine')}",
+      "32-#{_('Physics')}",
+      "12-#{_('Social and Behavioural Sciences')}",
+      "42-#{_('Thermal Engineering/Process Engineering')}"
```
I dont think the hashtag notation will work for this. In previous times we have used the format method so that gettext can identify the keys.  Translation gem is using gettext which uses the full string as a key in a hash object to look up the translations. so if a key is missing it wont translate.

So it should be something like ``_('23-Agriculture, Forestry, Horticulture and Veterinary Medicine')``.

Or you can try something like ```format("23-%{subject}", subject : _(Agriculture, Forestry, Horticulture and Veterinary Medicine'))``` which could make it more general. I am not 100% sure it works though.

Give it a try on the test project and talk with @200455939-yashu .

aaronskiba replied:
<img width="885" alt="Screenshot 2023-08-01 at 3 00 58 PM" src="https://github.com/portagenetwork/roadmap/assets/71047780/547a0d49-2159-4089-82e1-75ea5a38a3fe">
<img width="426" alt="Screenshot 2023-08-01 at 3 00 17 PM" src="https://github.com/portagenetwork/roadmap/assets/71047780/160607c9-1c20-4516-926e-2b21b1f65c06">

Cool, good to double check. I just tried creating a couple translations in dmp-assistant-test. I guess the current code format is ok for translation.io to pick up.

aaronskiba replied:
Actually, although it is getting picked up, this will still probably need to be fixed. When `.map do |subject|
      [subject.split('-').last, subject.gsub('-', ' ')]` is applied to the array, both elements in the array receive the translation. However, `subject.gsub('-', ' ')` should not be receiving the translation.

lagoan commented on app/views/research_outputs/metadata_standards/_search.html.erb:
```diff
@@ -7,8 +7,8 @@
              namespace: "metadata_standards",
-             label: "Metadata Standard",
-             search_examples: "(e.g. DataCite, Dublin, Biological, etc.)",
+             label: _("Metadata Standard"),
+             search_examples: _("(e.g. DataCite, Dublin, Biological, etc.)"),
```
Good catch! We should be taking note of these errors to see if we can share them through a PR on the latest version of roadmap.

lagoan commented on app/views/static_pages/about_us.html.erb:
```diff
@@ -34,10 +34,11 @@
-        <p><%= _("If you have an account please sign in and start creating or editing your DMP.") %></p>
+        <p><%= _("If you have an account, please sign in and start creating or editing your DMP.") %></p>
+        <!--
         <p><%= _("If you do not have a %{application_name} account, click on")  % {application_name: ApplicationService.application_name} %> <a href="<%= root_path %>"><%= _('Sign up') %></a> <%= _("on the homepage.") %></p>
```
We were using the ``format`` method in other instances, we should decide on an approach and stick with it. I am fine with both. Probably this simplified way is better though.

This is something else we can bring up in the dev meeting once we catch up to version 4.

aaronskiba replied:
I agree, it would be best to have a single approach throughout (issue created here https://github.com/portagenetwork/roadmap/issues/415). I think we originally tried this approach for the landing page fix, but Rubocop complained unless we used `format` (https://github.com/portagenetwork/roadmap/commit/a5c308a02540ff206964c710ec5d324601082ec5). So maybe `format` is best?

aaronskiba replied:
Also, I just noticed that one of these help links is broken on production https://github.com/portagenetwork/roadmap/issues/412.
For now, I just made a small edit to the existing code https://github.com/portagenetwork/roadmap/pull/409/commits/57f5d256b7ef409211c97bf72f86af3da122254c. But the reformatting you pointed out should be addressed soon as well.

lagoan replied:
We are using the simplified version on all our ERB files. Let's keep it as it is for now unless upstrem changes that.

lagoan commented on app/views/static_pages/about_us.html.erb:
```diff
@@ -34,10 +34,11 @@
 
-        <p><%= _("If you have an account please sign in and start creating or editing your DMP.") %></p>
+        <p><%= _("If you have an account, please sign in and start creating or editing your DMP.") %></p>
+        <!--
```
Sorry, why are we commenting out this next part? If we don't need it we should remove it, but it looks like this code is something we do want here.

aaronskiba replied:
All good. This related to #407. Currently, the "Sign Up" and "Help" messages are repeated in the "Getting Started" section of the "About Us" page https://assistant.portagenetwork.ca/about_us.

I'm thinking we'll need to remove one of the "Sign Up" messages and one of the "Help" messages. But I actually prefer the code I commented out in comparison to what I kept. However, translations aren't currently available for the code that I do prefer.

lagoan replied:
We can double-check with @isawil to see which copy we like best. We can leave just one instead of waiting for the translation for the missing piece.

Does this make sense?

aaronskiba replied:
Cool, yes, that makes sense. I'll verify which one we should keep.

lagoan reviewed (commented):
The changes are looking good! I just have a few questions about the approach for some minor parts of the code.

Looking forward to reading your responses!

aaronskiba commented:
> The changes are looking good! I just have a few questions about the approach for some minor parts of the code.
> 
> Looking forward to reading your responses!

Awesome, thank you! Translation.io is a little peculiar so I appreciate any questions/concerns.

It looks like all of the proposed changes are currently working for the translations, though. Maybe you'd still prefer for something to be refactored, though? Also, I don't know if you have a preference as to how we should proceed with this https://github.com/portagenetwork/roadmap/pull/409#discussion_r1281156978? Thanks!

aaronskiba committed:
Enable translation for 'Research domain' dropdown

aaronskiba committed:
Add comments to clarify translation.io sync

aaronskiba committed:
Prevent translation of subject.gsub('-', ' ')

Perhaps the best solution would be to add 'enum selectable_subject' to ResearchOutput. That way we could implement the same logic we used for ResearchOuput.output_types and ResearchOutput.accesses.

aaronskiba committed:
Fix broken help link & add to changelog

aaronskiba committed:
run `rake translation:sync` with prod key

200455939-yashu commented on config/locale/app.pot:
Do we need these files ?

aaronskiba replied:
Good question. I want to test the translations in UAT, but there are a few things I'm unsure about:
1) Does UAT use the test key or the prod key for translation.io?
2) Do I need to provide the po files for UAT deployment?
3) When translations are provided to translation.io, are those changes automatically sync'd back to UAT? And if not, should we deploy another UAT instance, or should we execute the `translation:sync` task within the existing UAT instance?

200455939-yashu replied:
As per my knowledge,
1. UAT uses the prod key.
2. No, PO files are not needed for the UAT deployment.
3. Everything available in Translation.io are directly available to the UAT.

aaronskiba commented:
> It looks like all of the proposed changes are currently working for the translations, though. Maybe you'd still prefer for something to be refactored, though? Also, I don't know if you have a preference as to how we should proceed with this [#409 (comment)](https://github.com/portagenetwork/roadmap/pull/409#discussion_r1281156978)? Thanks!

Or so I thought, haha. There was an issue that was addressed with the following commit: https://github.com/portagenetwork/roadmap/pull/409/commits/07741d216197351c4a9be74c9b11abf8435385ae

Also, sorry, I have a question here: https://github.com/portagenetwork/roadmap/pull/409#discussion_r1284637742

200455939-yashu reviewed (dismissed):
Good work

aaronskiba committed:
Revert "run `rake translation:sync` with prod key"

This reverts commit 1a00675b1afc58cab983e94d48894a1470438f09.

lagoan commented on app/presenters/research_output_presenter.rb:
```diff
@@ -11,13 +11,13 @@ def initialize(research_output:)
   def selectable_output_types
     ResearchOutput.output_types
-                  .map { |k, _v| [k.humanize, k] }
+                  .map { |k, _v| [_(k.humanize), k] } # k.humanize is sync'd to translation.io via _research_output.erb
```
I like how you are referencing where we need to look for the translations in case we ever need to update them.

lagoan commented on CHANGELOG.md:
```diff
@@ -10,6 +10,12 @@
+
+ - Removed repeated "Sign-up" and "Help" text in "About Us" [#407](https://github.com/portagenetwork/roadmap/issues/407)
+
+ - Fixed broken "Help" link [402](https://github.com/portagenetwork/roadmap/issues/412)
```
Minor change, add the # simbol to the link.

lagoan reviewed (changes requested):
Great work @aaronskiba , I found some very minor changes that we can make and then it will be good to go.

aaronskiba committed:
Remove commented out code

The commented out code is related to issue #407. It was decided that the commented code is what will be removed.

aaronskiba committed:
Add # symbol to link

aaronskiba commented:
Hi @lagoan, I did the changelog cleanup. Also, after discussing the repeated **About Us** text with @isawil, the commented-out code is what we will **not** be keeping. As a result, it has been removed.

lagoan reviewed (approved):
![LGTM](https://media.giphy.com/media/VhWVAa7rUtT3xKX6Cd/giphy.gif)
